### PR TITLE
ENH:read_html() handles tables with multiple header rows #13434

### LINKED ISSUE #13434: read_html() doesn't handle tables with multiple header rows
The `read_html()` function seems to treat every `<th>` in a table as a column, even if they occur in separate `<tr>`s. This means that it breaks even on simple tables generated by pandas' `to_html()` function.
#### Code Sample, a copy-pastable example if possible

```
df = pd.DataFrame(
    columns=["Name", "Age", "Party"], 
    data = [("Hillary", 68, "D"), ("Bernie", 74, "D"), ("Donald", 69, "R")])
df = df.set_index("Name")
html = df.to_html()
df2 = pd.read_html(html)[0]
print df2
```

This is the value of `html`, generated by the `to_html()` function on the original data frame:

<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Age</th>
      <th>Party</th>
    </tr>
    <tr>
      <th>Name</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Hillary</th>
      <td>68</td>
      <td>D</td>
    </tr>
    <tr>
      <th>Bernie</th>
      <td>74</td>
      <td>D</td>
    </tr>
    <tr>
      <th>Donald</th>
      <td>69</td>
      <td>R</td>
    </tr>
  </tbody>
</table>

```
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>Age</th>
      <th>Party</th>
    </tr>
    <tr>
      <th>Name</th>
      <th></th>
      <th></th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>Hillary</th>
      <td>68</td>
      <td>D</td>
    </tr>
...
  </tbody>
</table>
```

And this is the printed output of the newly-parsed dataframe `df2`:

```
  Unnamed: 0  Age Party  Name  Unnamed: 4  Unnamed: 5
0    Hillary   68     D   NaN         NaN         NaN
1     Bernie   74     D   NaN         NaN         NaN
2     Donald   69     R   NaN         NaN         NaN
```

What happens is that the `to_html()` function produces an html table with two header rows, one for the column names and one with the index name. However the `read_html()` parser interprets each individual `th` cell as an expected column, resulting in twice the number of columns. Even worse, this produces a column with the same name as the original index but without any data. 
#### Expected Output

The `read_html` parser could either treat the multi-row header fully correctly:

```
         Age Party
Name              
Hillary   68     D
Bernie    74     D
Donald    69     R
```

Or it could just ignore any rows after the first one:

```
  Unnamed: 0  Age Party
0    Hillary   68     D
1     Bernie   74     D
2     Donald   69     R
```
## INSTALLED VERSIONS

commit: None
python: 2.7.11.final.0
python-bits: 64
OS: Darwin
OS-release: 14.5.0
machine: x86_64
processor: i386
byteorder: little
LC_ALL: None
LANG: en_IE.UTF-8

pandas: 0.18.1
nose: 1.3.7
pip: 8.1.1
setuptools: 21.0.0
Cython: 0.23.4
numpy: 1.11.0
scipy: 0.17.0
statsmodels: 0.6.1
xarray: None
IPython: 4.1.1
sphinx: 1.3.5
patsy: 0.4.1
dateutil: 2.5.3
pytz: 2016.4
blosc: None
bottleneck: 1.0.0
tables: 3.2.2
numexpr: 2.5
matplotlib: 1.5.1
openpyxl: 2.1.2
xlrd: None
xlwt: None
xlsxwriter: None
lxml: 3.3.5
bs4: 4.4.1
html5lib: 1.0b3
httplib2: 0.9.2
apiclient: None
sqlalchemy: 1.0.11
pymysql: None
psycopg2: 2.6.1 (dt dec pq3 ext lo64)
jinja2: 2.8
boto: 2.40.0
pandas_datareader: None

### PR DESCRIPTION
**Pull request by brianhuey**
- [x] closes #13434
 - [x] 2 tests added / passed
 - [x] Multiple tr rows within a thead are now parsed, creating a multiindex header.

### CODE REVIEW

File: pandas/io/tests/test_html.py [hunk 1]
```diff
@@ -760,6 +760,17 @@ def test_keep_default_na(self):
         html_df = read_html(html_data, keep_default_na=True)[0]
         tm.assert_frame_equal(expected_df, html_df)
 
+    def test_multiple_header_rows(self):
+        expected_df = DataFrame(data=[("Hillary", 68, "D"),
```
Comment (jreback) [inline pandas/io/tests/test_html.py]: can you add the issue number as a comment here

File: pandas/io/tests/test_html.py [hunk 2]
```diff
@@ -869,6 +880,17 @@ def test_computer_sales_page(self):
         data = os.path.join(DATA_PATH, 'computer_sales_page.html')
         self.read_html(data, header=[0, 1])
 
+    def test_multiple_header_rows(self):
+        expected_df = DataFrame(data=[("Hillary", 68, "D"),
```
Comment (jreback) [inline pandas/io/tests/test_html.py]: same here

**Review by jreback — CHANGES_REQUESTED** [review]
pls add a note in whatsnew (enhancements for 0.20)

pls update doc-string in pd.read_html and docs in io.rst (mini-example)

is there a real-world (shorter) url that this can be tested on (with the network decorator)?

**Comment by brianhuey** [comment]
I looked for a good real-world url to write a test around, but a lot of the examples I could find seemed like they were from websites where the data may change over time, causing the test to fail. If you want I could add an html file to /io/tests/data and write something around that.

**Comment by jreback** [comment]
hmm, you are not specifying the header here, so this can *infer* automatically? (that its a multi-index)?

**Comment by brianhuey** [comment]
Yes, for instances where a `<table>` contains a `<thead>` and multiple `<tr>`, `_data_to_frame()` infers all header rows whereas currently it only infers the first row, omitting the rest from the dataframe.

**Comment by jreback** [comment]
@brianhuey ok, please make a note of this in doc-string and show an example in io.rst.

do we ever need to turn this off? (e.g. false positives)?

**Comment by brianhuey** [comment]
Currently there is no way to turn off the automatic parsing of `<thead>` elements in to column headers. Since this enhancement only makes that parsing more complete, I would say that adding a parameter might not be necessary. It could be manually turned off by converting the multiindex to a dataframe and concatenating it with the `read_html()` dataframe.

**Comment by jreback** [comment]
can you rebase (tests have moved).

@jorisvandenbossche @TomAugspurger comments?

**Comment by jreback** [comment]
@brianhuey looking at the original issue. This *does* work, though still lots of unnamed levels.

Is this expected?
```
In [1]: df = pd.DataFrame(
   ...:     columns=["Name", "Age", "Party"], 
   ...:     data = [("Hillary", 68, "D"), ("Bernie", 74, "D"), ("Donald", 69, "R")])
   ...: df = df.set_index("Name")
   ...: html = df.to_html()
   ...: 

In [2]: pd.read_html(html)[0]
Out[2]: 
  Unnamed: 0_level_0                Age              Party
                Name Unnamed: 1_level_1 Unnamed: 2_level_1
0            Hillary                 68                  D
1             Bernie                 74                  D
2             Donald                 69                  R

```

**Comment by brianhuey** [comment]
Yes, I could rename the unnamed levels to a blank string, but I was trying to maintain consistency with the original HTML parser which names blank column headers to 'Unnamed'. I assume it was done this way originally to remove any ambiguity amongst column names. I'm open to any suggestions on how to make this better.

**Comment by jreback** [comment]
@brianhuey I don't think thats right actually. 

```
In [2]: df
Out[2]: 
         Age Party
Name              
Hillary   68     D
Bernie    74     D
Donald    69     R
```
This is not a multi-line header. Rather ``Name`` is the index name. I *think* the ``to_html()`` is correct as well.

This is a case that you can actually detect (and we do in csv reading and such).
IOW If you have ALL empty string levels that you don't actually have a multi-line header.
In fact I would *make* the user specify it anyhow (IOW, ``headers=None``) means only a single header.

``headers=[0,1]`` is very explict and that is the ONLY way to create a multi-line header.

So I would say that you are actually reading a single line header, but with an index label. Which I can't believe is actually that common (except if we are generating it).

**Comment by brianhuey** [comment]
Correct, the test DataFrame has index 'Name' and cols 'Age' and 'Party'. When this is converted to HTML it becomes a table with headers that look something like this:
`<thead>`
`            <tr>`
`                  <th></th><th>Age</th><th>Party</th>`
`            </tr>`
`            <tr>`
`                   <th>Name</th><th></th><th></th>`
`            </tr>`
`</thead>`
Which is essentially the multi-line HTML table header described in [https://github.com/pandas-dev/pandas/issues/13434](url). You could imagine a similar table that looks like this:
`<thead>`
`            <tr>`
`                  <th></th><th>Age</th><th>Party</th><th>Party</th>`
`            </tr>`
`            <tr>`
`                   <th>Name</th><th></th><th></th><th></th>`
`            </tr>`
`</thead>`
My understanding is that selecting a specific 'Party' column becomes ambiguous unless you select by position. This was my rationale for using 'Unnamed' rather than making it look better by inserting blank strings.

**Comment by jreback** [comment]
@brianhuey can you post rendered versions of those as well (in the issue)

**Comment by brianhuey** [comment]
<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>Age</th>      <th>Party</th>    </tr>    <tr>      <th>Name</th>      <th></th>      <th></th>    </tr>  </thead>  <tbody>    <tr>      <th>Hillary</th>      <td>68</td>      <td>D</td>    </tr>    <tr>      <th>Bernie</th>      <td>74</td>      <td>D</td>    </tr>    <tr>      <th>Donald</th>      <td>69</td>      <td>R</td>    </tr>  </tbody></table>

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th></th>      <th>Age</th>      <th>Party</th>   <th>Party</th> </tr>    <tr>      <th>Name</th>      <th></th>      <th></th>  <th></th>  </tr>  </thead>  <tbody>    <tr>      <th>Hillary</th>      <td>68</td>      <td>D</td>   <td>B</td> </tr>    <tr>      <th>Bernie</th>      <td>74</td>      <td>D</td>   <td>C</td> </tr>    <tr>      <th>Donald</th>      <td>69</td>      <td>R</td>  <td>F</td>  </tr>  </tbody></table>

**Comment by jreback** [comment]
@brianhuey ok I think you can then treat this like we do in read_csv. Assuming that ``header`` is an int (or None), and NOT a list. If you find the pattern like you do above then you will have a named Index.

**Comment by brianhuey** [comment]
In the case of a multi-row HTML header, wouldn't we want to pass a list of header rows? For instance, in the function below, `head` contains our two HTML header rows: `[['','Age','Party'], ['Name','','']` These are appended to the dataframe, and then specified  (`header=[0,1]`) in TextParser. 
```
def _data_to_frame(**kwargs):
    head, body, foot = kwargs.pop('data')
    header = kwargs.pop('header')
    kwargs['skiprows'] = _get_skiprows(kwargs['skiprows'])
    if head:
        rows = lrange(len(head))
        body = head + body
        if header is None:  # special case when a table has <th> elements
            header = 0 if rows == [0] else rows

    if foot:
        body += [foot]

    # fill out elements of body that are "ragged"
    _expand_elements(body)
    tp = TextParser(body, header=header, **kwargs)
    df = tp.read()
    return df
```

**Comment by jreback** [comment]
@brianhuey 

I wouldn't expect to *need* to pass anything in this case for ``header`` (or ``header=0``) is equivalent. I think that it should/can infer that the ``Name`` row is actually an index label. The logic is not too complex and we do this in csv parsing. (in fact we accept *both* cases, IOW a named Index and no row at all between the header and the data).

This IMHO is NOT a multi-header at all.

**Comment by brianhuey** [comment]
But what about HTML tables where there are actually two rows of headers (i.e. two <tr> within a <thead> tag). The original issue raised was that pandas does not parse these multi-header row HTML tables correctly. If we are dealing with an HTML table like the one below, I would expect the parser would not automatically set the "Name" column to be the index, but I would expect the parser to identify the first two rows of the table as column headers and set those accordingly.

<table border="1" class="dataframe">  <thead>    <tr style="text-align: right;">      <th>Zero</th>      <th>Age</th>      <th>Party</th>    </tr>    <tr>      <th>Name</th>      <th>One</th>      <th>Two</th>    </tr>  </thead>  <tbody>    <tr>      <th>Hillary</th>      <td>68</td>      <td>D</td>    </tr>    <tr>      <th>Bernie</th>      <td>74</td>      <td>D</td>    </tr>    <tr>      <th>Donald</th>      <td>69</td>      <td>R</td>    </tr>  </tbody></table>

**Comment by jreback** [comment]
@brianhuey so for that one, I would say that you would have to specify ``header=[0,1]``

**Comment by jorisvandenbossche** [comment]
I am not sure if it is worth adding this complexity to read_html of inferring index names due to strange formatting of the headers. That is something pandas-specific, and `read_html` is not really regarded as something you store data to later read it again (as opposed to csv). 
I personally think that the last example of @brianhuey (two header levels) is indeed something that `read_html` could do automatically to set the two column rows as the MultiIndexed columns, as this is actually specified by the html source.

**Comment by jreback** [comment]
@brianhuey ok with @jorisvandenbossche suggestion. What we want is to have a simple / understandable approach that is the least magical as possible, but at the same time can get the job done.

**Comment by brianhuey** [comment]
I totally agree. The last table example is the one that I had in mind when I started out. Perhaps the test example confuses things as because indexing the 'Name' column is only done in order to generate a two-row HTML header when `df.to_html()` is called. I could easily add the last table example as the test case instead.

**Comment by AashitaK** [comment]
I am having hard time reading the table with multi-indexed columns from [census India](http://www.censusindia.gov.in/Census_Data_2001/Census_Data_Online/Language/parta.htm) using read_html. I just found the issue reported here and the discussion about how the original example reported was not multi-indexed, so thought it might be useful to mention this example.

**Comment by brianhuey** [comment]
@AashitaK the table you linked is a little more complex than the ones I'm trying to handle here as it has multiple column headers but they span a variable number of columns. I am interested in tackling this problem next.

**Comment by AashitaK** [comment]
@brianhuey Thanks.

**Comment by jreback** [comment]
@brianhuey ?

**Comment by brianhuey** [comment]
I believe the code achieves what @jorisvandenbossche suggests. It reads multiple header rows from HTML syntax, and parses them as a multiindex column, producing a dataframe with columns that match the structure of an HTML table. I'm open to any suggestions on where to improve how this is handled (i.e. the 'Unnamed: level' thing, specifying multiple headers).

File: pandas/tests/io/test_html.py [hunk 3]
```diff
@@ -869,6 +881,18 @@ def test_computer_sales_page(self):
         data = os.path.join(DATA_PATH, 'computer_sales_page.html')
         self.read_html(data, header=[0, 1])
 
+    def test_multiple_header_rows(self):
+        # Issue #13434
+        expected_df = DataFrame(data=[("Hillary", 68, "D"),
```
Comment (jreback) [inline pandas/tests/io/test_html.py]: is this test in here twice?

Comment (brianhuey) [inline pandas/tests/io/test_html.py]: Yes, it tests it twice: once using the BeautifulSoup parser and again with lxml.

Comment (jreback) [inline pandas/tests/io/test_html.py]: whoosh we should never actually duplicate the test code (not sure if this is existing)

rather use a base class (or parametization)

if this is the current style can u pls create an issue to fix this (if not pls fix this!)

Comment (brianhuey) [inline pandas/tests/io/test_html.py]: I see, I misunderstood what was going on there. I'll fix.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @brianhuey 

followups always welcome!

**Comment by brianhuey** [comment]
I appreciate you working with me through this. First time contributing to an open source project!

**Comment by jreback** [comment]
@brianhuey you did great! 

keep em coming!

**Pull request closed without merging**